pkg/nodejs/npm/parse.go

```diff
@@ -53,13 +61,106 @@ func (p *Parser) Parse(r dio.ReadSeekerAt) ([]types.Library, []types.Dependency,
 	}
 
-	dircetDeps := lockFile.Packages[""].Dependencies
-	libs, deps := p.parse(lockFile.Dependencies, dircetDeps, map[string]string{})
+	var libs []types.Library
+	var deps []types.Dependency
+	if lockFile.LockfileVersion == 1 {
+		libs, deps = p.parseV1(lockFile.Dependencies, map[string]string{})
+	} else {
+		libs, deps = p.parseV2(lockFile.Packages)
+	}
 
 	return utils.UniqueLibraries(libs), uniqueDeps(deps), nil
 }
 
-func (p *Parser) parse(dependencies map[string]Dependency, dircetDeps map[string]string, versions map[string]string) ([]types.Library, []types.Dependency) {
+func (p *Parser) parseV2(packages map[string]Package) ([]types.Library, []types.Dependency) {
+	libs := make(map[string]types.Library, len(packages)-1)
+	var deps []types.Dependency
+
+	directDeps := map[string]string{}
+	for name, version := range packages[""].Dependencies {
+		pkgPath := filepath.Join(nodeModulesFolder, name)
+		pkg, ok := packages[pkgPath]
+		if !ok {
+			log.Logger.Debugf("unable to find %s@%s", name, version)
+			continue
+		}
+		directDeps[name] = pkg.Version
+	}
+
+	for pkgPath, pkg := range packages {
+		if pkg.Dev || pkgPath == "" {
+			continue
+		}
+
+		pkgName := pkgNameFromPath(pkgPath)
+		pkgID := utils.PackageID(pkgName, pkg.Version)
+
+		// There are cases when similar libraries use same dependencies
+		// we need to add location for each these dependencies
+		if savedLib, ok := libs[pkgID]; ok {
+			savedLib.Locations = append(savedLib.Locations, types.Location{StartLine: pkg.StartLine, EndLine: pkg.EndLine})
+			libs[pkgID] = savedLib
+			continue
+		}
+
+		lib := types.Library{
+			ID:                 pkgID,
+			Name:               pkgName,
+			Version:            pkg.Version,
+			Indirect:           isIndirectLib(pkgName, pkg.Version, directDeps),
+			ExternalReferences: []types.ExternalRef{{Type: types.RefOther, URL: pkg.Resolved}},
+			Locations: []types.Location{
+				{
+					StartLine: pkg.StartLine,
+					EndLine:   pkg.EndLine,
+				},
+			},
+		}
+		libs[pkgID] = lib
+
+		dependsOn := make([]string, 0, len(pkg.Dependencies))
+		for depName, depVersion := range pkg.Dependencies {
+			// Try to resolve the version with nested dependencies first
+			depPath := filepath.Join(pkgPath, nodeModulesFolder, depName)
+			if dep, ok := packages[depPath]; ok {
+				depID := utils.PackageID(depName, dep.Version)
+				dependsOn = append(dependsOn, depID)
+				continue
+			}
+
+			// Try to resolve the version with dependencies same folder
+			depPath = filepath.Join(filepath.Dir(pkgPath), depName)
+			if dep, ok := packages[depPath]; ok {
+				depID := utils.PackageID(depName, dep.Version)
+				dependsOn = append(dependsOn, depID)
+				continue
+			}
+
+			// Try to resolve the version with the higher level dependencies
+			depPath = filepath.Join(nodeModulesFolder, depName)
+			if dep, ok := packages[depPath]; ok {
+				depID := utils.PackageID(depName, dep.Version)
+				dependsOn = append(dependsOn, depID)
+				continue
+			}
```

Comment:
Isn't it possible to have more than two nests like `node_modules/body-parser/node_modules/debug/node_modules/ms`? Looks like it happens theoretically.

What about this code?

```suggestion
			depPath := path.Join(pkgPath, nodeModulesFolder)
			paths := strings.Split(depPath, "/")

			// Try to resolve the version with the nearest directory
			// In the case of "ms", try the following paths in order.
			//    - "node_modules/body-parser/node_modules/debug/node_moduls/ms"
			//    - "node_modules/body-parser/node_modules/ms"
			//    - "node_modules/ms"
			for i := len(paths) - 1; i >= 0; i-- {
				if paths[i] != nodeModulesFolder {
					continue
				}
				p := path.Join(paths[:i+1]...)
				p = path.Join(p, depName)
	
				if dep, ok := packages[p]; ok {
					depID := utils.PackageID(depName, dep.Version)
					dependsOn = append(dependsOn, depID)
					continue
				}
			}
```

https://go.dev/play/p/UPQN11T8qJ_m

Reply:
Good idea! thanks, used this.